File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = INT_MIN;
    int maxPathSum(TreeNode* root) {
        
        dfs(root);
        return res;
    }
private: 
    int dfs(TreeNode* root) {
        if (root == NULL) return 0;
        
        int left = dfs(root -> left);
        int right = dfs(root -> right);
        
        int inLine = max(root-> val, max(left, right) + root -> val);
        int inRoot = max(inLine, root -> val + left + right);
        res = max(res, inRoot);
        
        return inLine;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        
        res = [float("-inf")]

        def dfs(node):

            if not node:
                return 0

            if not node.left and not node.right:
                res[0] = max(res[0], node.val)
                return node.val
            
            left_longest = max(0,dfs(node.left))
            right_longest = max(0, dfs(node.right))

            #  middle
            res[0] = max(res[0], node.val + left_longest + right_longest)

            #along the way
            return max(node.val + left_longest, node.val + right_longest)
        
        dfs(root)
        return res[0]